File: sdk/typescript/examples/basic-usage.ts
import { GotrsClient, isNotFoundError, isUnauthorizedError } from '../src';

// Example 1: Basic usage with API key
async function basicExample() {
  console.log('🚀 Basic GOTRS SDK Example');

  // Initialize client with API key
  const client = GotrsClient.withApiKey('https://your-gotrs-instance.com', 'your-api-key');

  try {
    // Test connection
    const isConnected = await client.ping();
    if (!isConnected) {
      throw new Error('Failed to connect to GOTRS API');
    }
    console.log('✅ Connected to GOTRS successfully');

    // Get dashboard stats
    const stats = await client.dashboard.getStats();
    console.log('📊 Dashboard Stats:', {
      totalTickets: stats.total_tickets,
      openTickets: stats.open_tickets,
      myTickets: stats.my_tickets,
    });

    // List recent tickets
    const tickets = await client.tickets.list({
      page_size: 5,
      sort_by: 'created_at',
      sort_order: 'desc',
    });
    
    console.log(`📋 Found ${tickets.total_count} tickets (showing ${tickets.tickets.length})`);
    tickets.tickets.forEach(ticket => {
      console.log(`  - #${ticket.ticket_number}: ${ticket.title} (${ticket.status})`);
    });

    // Create a new ticket
    const newTicket = await client.tickets.create({
      title: 'SDK Test Ticket',
      description: 'This ticket was created using the TypeScript SDK',
      priority: 'normal',
      type: 'incident',
      queue_id: 1,
      customer_id: 1,
      tags: ['sdk', 'test'],
    });
    
    console.log(`✅ Created ticket #${newTicket.ticket_number} with ID ${newTicket.id}`);

    // Add a message to the ticket
    const message = await client.tickets.addMessage(newTicket.id, {
      content: 'This is a test message added via the TypeScript SDK',
      message_type: 'note',
      is_internal: false,
    });
    
    console.log(`💬 Added message with ID ${message.id}`);

    // Update ticket priority
    const updatedTicket = await client.tickets.update(newTicket.id, {
      priority: 'high',
    });
    
    console.log(`🔄 Updated ticket priority to ${updatedTicket.priority}`);

    // Search tickets
    const searchResults = await client.tickets.search('SDK', {
      page_size: 3,
    });
    
    console.log(`🔍 Found ${searchResults.total_count} tickets matching 'SDK'`);

    // Close the test ticket
    const closedTicket = await client.tickets.close(newTicket.id, 'Test completed');
    console.log(`🔒 Closed ticket #${closedTicket.ticket_number}`);

  } catch (error) {
    console.error('❌ Error:', error);
  }
}

// Example 2: Authentication flow
async function authenticationExample() {
  console.log('\n🔐 Authentication Example');

  // Create client without authentication first
  const client = new GotrsClient({
    baseURL: 'https://your-gotrs-instance.com',
  });

  try {
    // Login with email and password
    await client.login('user@example.com', 'password');
    console.log('✅ Logged in successfully');

    // Now we can access protected resources
    const profile = await client.auth.getProfile();
    console.log(`👤 Logged in as: ${profile.first_name} ${profile.last_name} (${profile.email})`);

    // Get my tickets
    const myTickets = await client.dashboard.getMyTickets();
    console.log(`📝 I have ${myTickets.length} assigned tickets`);

    // Logout
    await client.logout();
    console.log('✅ Logged out successfully');

  } catch (error) {
    if (isUnauthorizedError(error)) {
      console.error('❌ Authentication failed: Invalid credentials');
    } else {
      console.error('❌ Error:', error);
    }
  }
}

// Example 3: Error handling
async function errorHandlingExample() {
  console.log('\n🚨 Error Handling Example');

  const client = GotrsClient.withApiKey('https://your-gotrs-instance.com', 'your-api-key');

  try {
    // Try to get a non-existent ticket
    await client.tickets.get(999999);
    
  } catch (error) {
    if (isNotFoundError(error)) {
      console.log('✅ Properly handled NotFound error');
    } else if (isUnauthorizedError(error)) {
      console.log('❌ Authentication error - check your API key');
    } else {
      console.error('❌ Unexpected error:', error);
    }
  }

  try {
    // Try to create an invalid ticket
    await client.tickets.create({
      title: '', // Invalid: empty title
      description: 'Test',
    });
    
  } catch (error) {
    console.log('✅ Properly handled validation error');
  }
}

// Example 4: Real-time events
async function realTimeExample() {
  console.log('\n⚡ Real-time Events Example');

  const client = GotrsClient.withApiKey('https://your-gotrs-instance.com', 'your-api-key');

  try {
    // Connect to real-time events
    await client.events.connect();
    console.log('✅ Connected to real-time events');

    // Listen for ticket events
    const ticketSubscription = client.events.onTicket((event) => {
      console.log(`🎫 Ticket event: ${event.type}`, {
        ticketNumber: event.data.ticket_number,
        title: event.data.title,
        status: event.data.status,
      });
    });

    // Listen for message events
    const messageSubscription = client.events.onMessage((event) => {
      console.log(`💬 Message event: ${event.type}`, {
        ticketId: event.data.ticket_id,
        content: event.data.content.substring(0, 50) + '...',
      });
    });

    // Listen for connection events
    const connectionSubscription = client.events.onConnection({
      connected: () => console.log('🔗 Connected to WebSocket'),
      disconnected: (event) => console.log('❌ Disconnected:', event.reason),
      error: (error) => console.error('❌ WebSocket error:', error),
    });

    // Simulate some activity (in real app, this would happen naturally)
    console.log('⏳ Listening for events for 30 seconds...');
    
    setTimeout(() => {
      console.log('🛑 Stopping event listeners');
      ticketSubscription.unsubscribe();
      messageSubscription.unsubscribe();
      connectionSubscription.unsubscribe();
      client.events.disconnect();
    }, 30000);

  } catch (error) {
    console.error('❌ Real-time events error:', error);
  }
}

// Example 5: Concurrent operations
async function concurrentExample() {
  console.log('\n🚀 Concurrent Operations Example');

  const client = GotrsClient.withApiKey('https://your-gotrs-instance.com', 'your-api-key');

  try {
    // Perform multiple operations concurrently
    const [stats, tickets, users, queues] = await Promise.all([
      client.dashboard.getStats(),
      client.tickets.list({ page_size: 5 }),
      client.users.list(),
      client.queues.list(),
    ]);

    console.log('✅ Fetched data concurrently:');
    console.log(`  - Dashboard stats: ${stats.total_tickets} total tickets`);
    console.log(`  - Recent tickets: ${tickets.tickets.length} tickets`);
    console.log(`  - Users: ${users.length} users`);
    console.log(`  - Queues: ${queues.length} queues`);

    // Create multiple tickets concurrently
    const ticketPromises = Array.from({ length: 3 }, (_, i) =>
      client.tickets.create({
        title: `Concurrent Ticket ${i + 1}`,
        description: `This is ticket ${i + 1} created concurrently`,
        priority: 'normal',
        queue_id: 1,
        customer_id: 1,
      })
    );

    const createdTickets = await Promise.all(ticketPromises);
    console.log(`✅ Created ${createdTickets.length} tickets concurrently`);

    createdTickets.forEach((ticket, index) => {
      console.log(`  - Ticket ${index + 1}: #${ticket.ticket_number}`);
    });

  } catch (error) {
    console.error('❌ Concurrent operations error:', error);
  }
}

// Example 6: File upload
async function fileUploadExample() {
  console.log('\n📎 File Upload Example');

  const client = GotrsClient.withApiKey('https://your-gotrs-instance.com', 'your-api-key');

  try {
    // First create a ticket
    const ticket = await client.tickets.create({
      title: 'Ticket with Attachment',
      description: 'This ticket will have a file attachment',
      queue_id: 1,
      customer_id: 1,
    });

    // Create a simple text file (in Node.js environment)
    if (typeof Buffer !== 'undefined') {
      const fileContent = 'This is a test file attachment from the TypeScript SDK';
      const fileBuffer = Buffer.from(fileContent, 'utf-8');

      const attachment = await client.tickets.uploadAttachment(
        ticket.id,
        fileBuffer,
        'test-file.txt'
      );

      console.log(`✅ Uploaded attachment: ${attachment.filename} (${attachment.size} bytes)`);

      // List all attachments for the ticket
      const attachments = await client.tickets.getAttachments(ticket.id);
      console.log(`📎 Ticket has ${attachments.length} attachment(s)`);

      // Download the attachment
      const downloadedData = await client.tickets.downloadAttachment(ticket.id, attachment.id);
      console.log(`⬇️ Downloaded ${downloadedData.byteLength} bytes`);
    } else {
      console.log('ℹ️ File upload example requires Node.js Buffer support');
    }

  } catch (error) {
    console.error('❌ File upload error:', error);
  }
}

// Run all examples
async function runAllExamples() {
  await basicExample();
  await authenticationExample();
  await errorHandlingExample();
  await realTimeExample();
  await concurrentExample();
  await fileUploadExample();
  
  console.log('\n✨ All examples completed!');
}

// Export examples for individual testing
export {
  basicExample,
  authenticationExample,
  errorHandlingExample,
  realTimeExample,
  concurrentExample,
  fileUploadExample,
  runAllExamples,
};

// Run examples if this file is executed directly
if (require.main === module) {
  runAllExamples().catch(console.error);
}

File: sdk/go/examples/basic_usage.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/gotrs-io/gotrs-ce/sdk/go"
	"github.com/gotrs-io/gotrs-ce/sdk/go/types"
)

func main() {
	// Initialize client with API key
	client := gotrs.NewClientWithAPIKey("https://your-gotrs-instance.com", "your-api-key")

	ctx := context.Background()

	// Test connection
	if err := client.Ping(ctx); err != nil {
		log.Fatalf("Failed to connect to GOTRS: %v", err)
	}
	fmt.Println("✅ Connected to GOTRS successfully")

	// List tickets
	fmt.Println("\n📋 Listing tickets...")
	tickets, err := client.Tickets.List(ctx, &types.TicketListOptions{
		PageSize: 10,
		Status:   []string{"open", "pending"},
	})
	if err != nil {
		log.Fatalf("Failed to list tickets: %v", err)
	}
	fmt.Printf("Found %d tickets\n", tickets.TotalCount)

	for _, ticket := range tickets.Tickets {
		fmt.Printf("- #%s: %s (Status: %s, Priority: %s)\n",
			ticket.TicketNumber, ticket.Title, ticket.Status, ticket.Priority)
	}

	// Create a new ticket
	fmt.Println("\n🎫 Creating a new ticket...")
	newTicket, err := client.Tickets.Create(ctx, &types.TicketCreateRequest{
		Title:       "SDK Test Ticket",
		Description: "This ticket was created using the Go SDK",
		Priority:    "normal",
		Type:        "incident",
		QueueID:     1,
		CustomerID:  1,
		Tags:        []string{"sdk", "test"},
	})
	if err != nil {
		log.Fatalf("Failed to create ticket: %v", err)
	}
	fmt.Printf("✅ Created ticket #%s with ID %d\n", newTicket.TicketNumber, newTicket.ID)

	// Add a message to the ticket
	fmt.Println("\n💬 Adding a message...")
	message, err := client.Tickets.AddMessage(ctx, newTicket.ID, &types.MessageCreateRequest{
		Content:     "This is a test message added via the SDK",
		MessageType: "note",
		IsInternal:  false,
	})
	if err != nil {
		log.Fatalf("Failed to add message: %v", err)
	}
	fmt.Printf("✅ Added message with ID %d\n", message.ID)

	// Update ticket priority
	fmt.Println("\n🔄 Updating ticket priority...")
	priority := "high"
	updatedTicket, err := client.Tickets.Update(ctx, newTicket.ID, &types.TicketUpdateRequest{
		Priority: &priority,
	})
	if err != nil {
		log.Fatalf("Failed to update ticket: %v", err)
	}
	fmt.Printf("✅ Updated ticket priority to %s\n", updatedTicket.Priority)

	// Assign ticket to user
	fmt.Println("\n👤 Assigning ticket...")
	assignedTicket, err := client.Tickets.Assign(ctx, newTicket.ID, 1) // Assign to user ID 1
	if err != nil {
		log.Fatalf("Failed to assign ticket: %v", err)
	}
	fmt.Printf("✅ Assigned ticket to user ID %d\n", *assignedTicket.AssignedTo)

	// Search tickets
	fmt.Println("\n🔍 Searching tickets...")
	searchResults, err := client.Tickets.Search(ctx, "SDK", &types.TicketListOptions{
		PageSize: 5,
	})
	if err != nil {
		log.Fatalf("Failed to search tickets: %v", err)
	}
	fmt.Printf("Found %d tickets matching 'SDK'\n", searchResults.TotalCount)

	// Get dashboard stats
	fmt.Println("\n📊 Getting dashboard statistics...")
	stats, err := client.Dashboard.GetStats(ctx)
	if err != nil {
		log.Fatalf("Failed to get dashboard stats: %v", err)
	}
	fmt.Printf("📈 Dashboard Stats:\n")
	fmt.Printf("  Total Tickets: %d\n", stats.TotalTickets)
	fmt.Printf("  Open Tickets: %d\n", stats.OpenTickets)
	fmt.Printf("  Closed Tickets: %d\n", stats.ClosedTickets)
	fmt.Printf("  My Tickets: %d\n", stats.MyTickets)

	// Get user profile
	fmt.Println("\n👤 Getting user profile...")
	profile, err := client.Auth.GetProfile(ctx)
	if err != nil {
		log.Fatalf("Failed to get profile: %v", err)
	}
	fmt.Printf("Logged in as: %s %s (%s)\n", profile.FirstName, profile.LastName, profile.Email)

	// Close the test ticket
	fmt.Println("\n🔒 Closing test ticket...")
	closedTicket, err := client.Tickets.Close(ctx, newTicket.ID, "Test completed")
	if err != nil {
		log.Fatalf("Failed to close ticket: %v", err)
	}
	fmt.Printf("✅ Closed ticket #%s\n", closedTicket.TicketNumber)

	fmt.Println("\n✨ SDK example completed successfully!")
}

// Example of error handling
func handleTicketOperations(client *gotrs.Client) {
	ctx := context.Background()

	// Example with proper error handling
	ticket, err := client.Tickets.Get(ctx, 12345)
	if err != nil {
		if gotrs.IsNotFound(err) {
			fmt.Println("Ticket not found")
			return
		}
		if gotrs.IsUnauthorized(err) {
			fmt.Println("Authentication failed")
			return
		}
		if gotrs.IsRateLimited(err) {
			fmt.Println("Rate limit exceeded, retrying later...")
			time.Sleep(time.Minute)
			return
		}
		log.Fatalf("Unexpected error: %v", err)
	}

	fmt.Printf("Found ticket: %s\n", ticket.Title)
}

// Example of using different authentication methods
func authenticationExamples() {
	// API Key authentication
	client1 := gotrs.NewClientWithAPIKey("https://gotrs.example.com", "your-api-key")

	// JWT authentication
	expiresAt := time.Now().Add(24 * time.Hour)
	client2 := gotrs.NewClientWithJWT("https://gotrs.example.com", "jwt-token", "refresh-token", expiresAt)

	// OAuth2 authentication
	oauth2Auth := gotrs.NewOAuth2Auth("access-token", "refresh-token", "Bearer", expiresAt)
	client3 := gotrs.NewClient(&gotrs.Config{
		BaseURL: "https://gotrs.example.com",
		Auth:    oauth2Auth,
		Timeout: 30 * time.Second,
		Debug:   true,
	})

	// Use the clients...
	_ = client1
	_ = client2
	_ = client3
}

// Example of concurrent operations
func concurrentOperations(client *gotrs.Client) {
	ctx := context.Background()

	// Create multiple tickets concurrently
	type ticketResult struct {
		ticket *types.Ticket
		err    error
	}

	results := make(chan ticketResult, 5)

	for i := 0; i < 5; i++ {
		go func(index int) {
			ticket, err := client.Tickets.Create(ctx, &types.TicketCreateRequest{
				Title:       fmt.Sprintf("Concurrent Ticket %d", index),
				Description: fmt.Sprintf("Created concurrently #%d", index),
				Priority:    "normal",
				QueueID:     1,
				CustomerID:  1,
			})
			results <- ticketResult{ticket: ticket, err: err}
		}(i)
	}

	// Collect results
	for i := 0; i < 5; i++ {
		result := <-results
		if result.err != nil {
			fmt.Printf("Failed to create ticket: %v\n", result.err)
		} else {
			fmt.Printf("Created ticket #%s\n", result.ticket.TicketNumber)
		}
	}
}
